File: src/main/java/com/truveta/opentoken/io/parquet/PersonAttributesParquetReader.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.parquet;

import org.apache.parquet.hadoop.ParquetReader;
import org.apache.parquet.schema.GroupType;
import org.apache.parquet.schema.Type;
import org.apache.parquet.example.data.Group;
import org.apache.parquet.hadoop.example.GroupReadSupport;
import org.apache.hadoop.conf.Configuration;

import com.truveta.opentoken.attributes.Attribute;
import com.truveta.opentoken.attributes.AttributeLoader;
import com.truveta.opentoken.io.PersonAttributesReader;

import org.apache.hadoop.fs.Path;

import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Set;
import java.io.Closeable;

/**
 * Reads person attributes from a Parquet file.
 * Implements the {@link PersonAttributesReader} interface.
 */
public class PersonAttributesParquetReader implements PersonAttributesReader, Closeable {
    private ParquetReader<Group> reader;
    private Group currentGroup;
    private Iterator<Group> iterator;
    private boolean closed = false;
    private boolean hasNextCalled = false;

    private Map<String, Attribute> attributeMap = new HashMap<>();

    /**
     * Initialize the class with the input file in Parquet format.
     * 
     * @param filePath the input file path
     * @throws IOException if an I/O error occurs
     */
    public PersonAttributesParquetReader(String filePath) throws IOException {
        Configuration conf = new Configuration();
        Path path = new Path(filePath);

        Set<Attribute> attributes = AttributeLoader.load();
        for (Attribute attribute : attributes) {
            for (String alias : attribute.getAliases()) {
                attributeMap.put(alias.toLowerCase(), attribute);
            }

        }

        GroupReadSupport readSupport = new GroupReadSupport();

        this.reader = ParquetReader.builder(readSupport, path).withConf(conf).build();

        this.iterator = new Iterator<Group>() {
            @Override
            public boolean hasNext() {
                try {
                    currentGroup = reader.read();
                    return currentGroup != null;
                } catch (IOException e) {
                    return false;
                }
            }

            @Override
            public Group next() {
                return currentGroup;
            }
        };
    }

    @Override
    public boolean hasNext() {
        if (closed) {
            throw new NoSuchElementException("Reader is closed");
        }
        if (!hasNextCalled) {
            hasNextCalled = iterator.hasNext();
        }
        return hasNextCalled;
    }

    @Override
    public Map<Class<? extends Attribute>, String> next() {
        if (closed || !hasNextCalled) {
            throw new NoSuchElementException("Reader is closed");
        }
        hasNextCalled = false;

        Group group = iterator.next();
        Map<Class<? extends Attribute>, String> attributes = new HashMap<>();
        GroupType schema = group.getType();
        for (Type field : schema.getFields()) {
            String fieldName = field.getName();
            int fieldIndex = schema.getFieldIndex(fieldName);
            if (group.getFieldRepetitionCount(fieldIndex) > 0) {
                String fieldValue = group.getValueToString(fieldIndex, 0);
                attributes.put(attributeMap.get(fieldName.toLowerCase()).getClass(), fieldValue);
            }
        }

        return attributes;
    }

    @Override
    public void close() throws IOException {
        reader.close();
        closed = true;
    }
}

File: lib/nodejs/opentoken/src/io/parquet/PersonAttributesParquetReader.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { ParquetReader } from 'parquetjs';
import { ParquetCursor } from 'parquetjs/lib/reader';
import { Attribute } from '../../attributes/Attribute';
import { AttributeLoader } from '../../attributes/AttributeLoader';
import { PersonAttributesReader } from '../PersonAttributesReader';

/**
 * Reads person attributes from a Parquet file.
 * Implements the PersonAttributesReader interface.
 */
export class PersonAttributesParquetReader implements PersonAttributesReader {
  private reader: ParquetReader | null = null;
  private cursor: ParquetCursor | null = null;
  private currentRecord: any = null;
  private closed = false;
  private hasNextCalled = false;
  private hasNextResult = false;
  private attributeMap: Map<string, Attribute> = new Map();

  /**
   * Initialize the class with the input file in Parquet format.
   *
   * @param filePath the input file path
   */
  constructor(private filePath: string) {}

  /**
   * Initialize the reader (async operation)
   */
  async initialize(): Promise<void> {
    // Load attributes and build the mapping
    const attributes = AttributeLoader.load();
    for (const attribute of attributes) {
      for (const alias of attribute.getAliases()) {
        this.attributeMap.set(alias.toLowerCase(), attribute);
      }
    }

    // Open the Parquet file
    this.reader = await ParquetReader.openFile(this.filePath);
    this.cursor = this.reader.getCursor();
  }

  /**
   * Check if there are more records to read.
   *
   * @returns true if there are more records, false otherwise
   * @throws Error if the reader is closed
   */
  async hasNext(): Promise<boolean> {
    if (this.closed) {
      throw new Error('Reader is closed');
    }

    if (!this.hasNextCalled) {
      if (!this.cursor) {
        throw new Error('Reader not initialized. Call initialize() first.');
      }
      this.currentRecord = await this.cursor.next();
      this.hasNextResult = this.currentRecord !== null;
      this.hasNextCalled = true;
    }

    return this.hasNextResult;
  }

  /**
   * Get the next record from the Parquet file.
   *
   * @returns a person attributes map
   * @throws Error when there are no more records or reader is closed
   */
  async next(): Promise<Map<string, string>> {
    if (this.closed || !this.hasNextCalled) {
      throw new Error('Reader is closed or hasNext() not called');
    }

    if (!this.hasNextResult) {
      throw new Error('No more records');
    }

    this.hasNextCalled = false;

    // Map to attribute classes
    const attributes = new Map<string, string>();

    for (const [fieldName, fieldValue] of Object.entries(this.currentRecord)) {
      const attribute = this.attributeMap.get(fieldName.toLowerCase());
      if (attribute) {
        const attributeName = attribute.getName();
        const fieldValueStr =
          fieldValue !== null && fieldValue !== undefined ? String(fieldValue) : '';
        attributes.set(attributeName, fieldValueStr);
      }
    }

    return attributes;
  }

  /**
   * Close the Parquet reader and release resources.
   */
  async close(): Promise<void> {
    if (this.reader) {
      await this.reader.close();
    }
    this.closed = true;
  }
}
